lib/Dialect/Polynomial/Conversions/PolynomialToModArith/PolynomialToModArith.cpp

```diff
@@ -0,0 +1,456 @@
+
+  LogicalResult matchAndRewrite(
+      ToTensorOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOp(op, adaptor.getOperands()[0].getDefiningOp());
+    return success();
+  }
+};
+
+struct ConvertConstant : public OpConversionPattern<ConstantOp> {
+  ConvertConstant(mlir::MLIRContext *context)
+      : OpConversionPattern<ConstantOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      ConstantOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ConvertCommon<ConstantOp> c(op, adaptor, typeConverter);
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto intAttr = dyn_cast<TypedIntPolynomialAttr>(op.getValue());
+    auto floatAttr = dyn_cast<TypedFloatPolynomialAttr>(op.getValue());
+    if (!intAttr && !floatAttr) return failure();
+
+    SmallVector<Attribute> coeffs;
+    coeffs.reserve(c.resultShape);
+    // This is inefficient for large-degree polys, but as of this writing we
+    // don't have a lowering that uses a sparse representation.
+    for (size_t i = 0; i < c.resultShape; ++i) {
+      if (intAttr) {
+        coeffs.push_back(rewriter.getIntegerAttr(c.resultElementType, 0));
+      } else {
+        coeffs.push_back(rewriter.getFloatAttr(c.resultElementType, 0.0));
+      }
+    }
+
+    // WARNING: if you don't store the IntPolynomial as an intermediate value
+    // before iterating over the terms, you will get a user-after-free bug.
+    // See the "Temporary range expression" section in
+    // https://en.cppreference.com/w/cpp/language/range-for
+    if (intAttr) {
+      const IntPolynomial &poly = intAttr.getValue().getPolynomial();
+      for (const auto &term : poly.getTerms()) {
+        int64_t idx = term.getExponent().getSExtValue();
+        APInt coeff = term.getCoefficient().sextOrTrunc(c.resultElementWidth);
+        coeffs[idx] = rewriter.getIntegerAttr(c.resultElementType, coeff);
+      }
+    } else {
+      const FloatPolynomial &poly = floatAttr.getValue().getPolynomial();
+      for (const auto &term : poly.getTerms()) {
+        int64_t idx = term.getExponent().getSExtValue();
+        // FIXME: only supports f64 now as TypedFloatPolynomialAttr only
+        // parses f64 inputs
+        APFloat coeff = term.getCoefficient();
+        coeffs[idx] = rewriter.getFloatAttr(c.resultElementType, coeff);
+      }
+    }
+
+    Value constantOp = b.create<arith::ConstantOp>(
+        DenseElementsAttr::get(c.resultTensorType, coeffs));
+
+    constantOp = modArithReduceOp(c, b, rewriter, constantOp, c.resultShape);
+
+    rewriter.replaceOp(op, constantOp);
+    return success();
+  }
+};
+
+template <typename Op, typename ModArithOp, typename ArithIOp,
+          typename ArithFOp>
+Value modArithBinaryOp(ConvertCommon<Op> &c, ImplicitLocOpBuilder &b, Value lhs,
+                       Value rhs) {
+  if (c.coefficientModulusIsNotNatural) {
```

Comment:
I think it would be best if we didn't even need to check this in this lowering and it should just always lower to `mod_arith.[add|sub|mul]`. Then we can check all that stuff there. As part of that we should also extend the `mod_arith` ops to include handling of floating point types. I think this is the same as what @AlexanderViand-Intel  is saying. So we could leave this as is and then file the following:

    - Extend `mod_arith` ops to support floating-point values as well.
    - Move this logic to the `mod_arith` to `arith` lowering.

Reply:
I have now migrated the logic of natural modulus handling inside `mod_arith`. I agree with the idea of handling all the integer stuff in `mod_arith`. But I think handling float point inside `mod_arith` seems a violation of its name.

Also, the `PolynomialToModArith` code must contain some float point specific thing, like `ConvertConstant` where `FloatAttr` is explicitly asked.